c10/util/ArrayRef.h

```diff
@@ -22,10 +22,87 @@
 
 #include <array>
 #include <iterator>
+#include <memory>
+#include <type_traits>
 #include <vector>
 
 namespace c10 {
 
+
+
+struct TORCH_API SymbolicIntImpl {
+  bool virtual isSymbolicInt(int64_t data_) = 0;
+
+  // I can't overload + for primitive types
+  int64_t virtual add(int64_t s1, int64_t s2) = 0;
+  bool virtual equal(int64_t s1, int64_t s2) = 0;
+};
+
+
+class TORCH_API SymbolicOrConcreteInt {
+
+public:
+
+  SymbolicOrConcreteInt(int64_t d):
+  data_(d) {};
+
+  operator int64_t() {
+    return data_;
+  }
+
+  int64_t data_;
+
+  // Note this is static as it should apply to all SymbolicInts
+  // we also want this to be fast enough, hence the raw ptr
+  static SymbolicIntImpl* impl_;
+
+
+  // TODO: this can be done per device
+  static void RegisterSymbolicIntImpl(SymbolicIntImpl* impl) {
+    impl_ = impl;
+  }
+
+  bool isSymbolicInt() {
+    return impl_->isSymbolicInt(data_);
+  }
+
+  bool operator==(const SymbolicOrConcreteInt& p2)
+  {
+      return SymbolicOrConcreteInt::impl_->equal(this->data_, p2.data_);
+  }
+
+  SymbolicOrConcreteInt operator+(SymbolicOrConcreteInt sci) {
+
+    // TODO: is there a way to force a conversion to "SymbolicInt" w/e it is
+    // before we start tracing ops on it?
+    // Otherwise this might not be a bad compromise to make this union to double back
+    // as a proxy object for a symbolic int
+
+    if (this->isSymbolicInt() || sci.isSymbolicInt()) {
+          int64_t data = impl_->add(data_, sci.data_);
+          return SymbolicOrConcreteInt(data);
+    } else {
+      return data_ + sci.data_;
+    }
+  }
+};
+
+
+
+class SymbolicInt {
+
+public:
+
+  SymbolicInt(int64_t d):
+  data_(d) {};
+
+  operator int64_t() {
+    return data_;
+  }
```

Comment:
No, we shouldn't have this implicit conversion

Reply:
It's unchecked! So you can end up with int64_t which represents a symbolic int (bad bad)